File: backend/src/main/java/com/example/backend/service/CollaborationService.java
package com.example.backend.service;

import com.example.backend.dto.*;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.Set;
import java.util.HashSet;
import java.util.stream.Collectors;

@Service
public class CollaborationService {

    private final SimpMessagingTemplate messagingTemplate;
    private final Map<Long, Set<String>> dossierViewers = new ConcurrentHashMap<>();
    private final Map<Long, Integer> dossierVersions = new ConcurrentHashMap<>();
    private final Map<String, String> userColors = new ConcurrentHashMap<>();
    private final String[] colorPalette = {
        "#FF5733", "#33FF57", "#3357FF", "#FF33A1", "#FFC300",
        "#DAF7A6", "#C70039", "#900C3F", "#581845", "#2ECC71"
    };
    private int colorIndex = 0;

    public CollaborationService(SimpMessagingTemplate messagingTemplate) {
        this.messagingTemplate = messagingTemplate;
    }

    public void handleUserJoined(Long dossierId, String userId, String username) {
        dossierViewers.computeIfAbsent(dossierId, k -> ConcurrentHashMap.newKeySet()).add(userId);
        
        if (!userColors.containsKey(userId)) {
            userColors.put(userId, colorPalette[colorIndex % colorPalette.length]);
            colorIndex++;
        }

        CollaborationPresenceDto presence = new CollaborationPresenceDto(userId, username, dossierId, "joined");
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/presence", presence);
        
        sendCurrentViewers(dossierId);
    }

    public void handleUserLeft(Long dossierId, String userId, String username) {
        Set<String> viewers = dossierViewers.get(dossierId);
        if (viewers != null) {
            viewers.remove(userId);
            if (viewers.isEmpty()) {
                dossierViewers.remove(dossierId);
            }
        }

        CollaborationPresenceDto presence = new CollaborationPresenceDto(userId, username, dossierId, "left");
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/presence", presence);
        
        sendCurrentViewers(dossierId);
    }

    public void broadcastCursor(Long dossierId, String userId, String username, String fieldName, Integer cursorPosition) {
        String color = userColors.getOrDefault(userId, "#000000");
        CollaborationCursorDto cursor = new CollaborationCursorDto(userId, username, dossierId, fieldName, cursorPosition, color);
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/cursor", cursor);
    }

    public CollaborationEditDto broadcastEdit(Long dossierId, String userId, String username, String fieldName, Object newValue, Object oldValue) {
        Integer currentVersion = dossierVersions.getOrDefault(dossierId, 0);
        Integer newVersion = currentVersion + 1;
        dossierVersions.put(dossierId, newVersion);

        CollaborationEditDto edit = new CollaborationEditDto(userId, username, dossierId, fieldName, newValue, oldValue, newVersion);
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/edit", edit);
        
        return edit;
    }

    public void broadcastActivity(Long dossierId, String userId, String username, String activityType, String description, Object data) {
        CollaborationActivityDto activity = new CollaborationActivityDto(userId, username, dossierId, activityType, description, data);
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/activity", activity);
    }

    public void shareFilterPreset(SharedFilterPresetDto filterPreset) {
        if (filterPreset.getSharedWithUserIds() != null) {
            for (String userId : filterPreset.getSharedWithUserIds()) {
                messagingTemplate.convertAndSendToUser(userId, "/queue/filter-presets", filterPreset);
            }
        } else {
            messagingTemplate.convertAndSend("/topic/filter-presets", filterPreset);
        }
    }

    public Set<String> getCurrentViewers(Long dossierId) {
        return dossierViewers.getOrDefault(dossierId, new HashSet<>());
    }

    public Integer getCurrentVersion(Long dossierId) {
        return dossierVersions.getOrDefault(dossierId, 0);
    }

    public String getUserColor(String userId) {
        return userColors.getOrDefault(userId, "#000000");
    }

    private void sendCurrentViewers(Long dossierId) {
        Set<String> viewers = getCurrentViewers(dossierId);
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/viewers", viewers);
    }

    public void broadcastConflictResolution(Long dossierId, String fieldName, Object resolvedValue, String resolvedBy) {
        Map<String, Object> conflictData = Map.of(
            "fieldName", fieldName,
            "resolvedValue", resolvedValue,
            "resolvedBy", resolvedBy,
            "version", getCurrentVersion(dossierId)
        );
        messagingTemplate.convertAndSend("/topic/dossier/" + dossierId + "/conflict", conflictData);
    }
}

File: frontend/src/app/services/collaboration.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { Observable, Subject, BehaviorSubject } from 'rxjs';
import { WebSocketService } from './websocket.service';
import { HttpClient } from '@angular/common/http';

export interface CollaborationPresence {
  userId: string;
  username: string;
  dossierId: number;
  action: 'joined' | 'left';
  timestamp: Date;
}

export interface CollaborationCursor {
  userId: string;
  username: string;
  dossierId: number;
  fieldName: string;
  cursorPosition: number;
  color: string;
  timestamp: Date;
}

export interface CollaborationEdit {
  userId: string;
  username: string;
  dossierId: number;
  fieldName: string;
  newValue: any;
  oldValue: any;
  version: number;
  editId: string;
  timestamp: Date;
  metadata?: any;
}

export interface CollaborationActivity {
  userId: string;
  username: string;
  dossierId: number;
  activityType: string;
  description: string;
  data: any;
  timestamp: Date;
}

export interface SharedFilterPreset {
  id: number;
  name: string;
  description: string;
  filters: any;
  sharedBy: string;
  sharedByUsername: string;
  sharedAt: Date;
  sharedWithUserIds?: string[];
}

export interface ConflictResolution {
  fieldName: string;
  resolvedValue: any;
  resolvedBy: string;
  version: number;
}

@Injectable({
  providedIn: 'root'
})
export class CollaborationService implements OnDestroy {
  private currentDossierId: number | null = null;
  private currentUserId: string | null = null;
  private currentUsername: string | null = null;

  private viewers$ = new BehaviorSubject<Set<string>>(new Set());
  private presence$ = new Subject<CollaborationPresence>();
  private cursor$ = new Subject<CollaborationCursor>();
  private edit$ = new Subject<CollaborationEdit>();
  private activity$ = new Subject<CollaborationActivity>();
  private conflict$ = new Subject<ConflictResolution>();
  private filterPreset$ = new Subject<SharedFilterPreset>();

  private pendingEdits: Map<string, CollaborationEdit> = new Map();
  private fieldVersions: Map<string, number> = new Map();

  constructor(
    private webSocketService: WebSocketService,
    private http: HttpClient
  ) {}

  async initializeForDossier(dossierId: number, userId: string, username: string): Promise<void> {
    this.currentDossierId = dossierId;
    this.currentUserId = userId;
    this.currentUsername = username;

    await this.webSocketService.connect(userId, username);

    this.subscribeToChannels(dossierId);
    this.joinDossier(dossierId, userId, username);
  }

  private subscribeToChannels(dossierId: number): void {
    this.webSocketService.subscribe<CollaborationPresence>(`/topic/dossier/${dossierId}/presence`)
      .subscribe(presence => {
        this.presence$.next(presence);
      });

    this.webSocketService.subscribe<Set<string>>(`/topic/dossier/${dossierId}/viewers`)
      .subscribe(viewers => {
        this.viewers$.next(new Set(viewers));
      });

    this.webSocketService.subscribe<CollaborationCursor>(`/topic/dossier/${dossierId}/cursor`)
      .subscribe(cursor => {
        if (cursor.userId !== this.currentUserId) {
          this.cursor$.next(cursor);
        }
      });

    this.webSocketService.subscribe<CollaborationEdit>(`/topic/dossier/${dossierId}/edit`)
      .subscribe(edit => {
        if (edit.userId !== this.currentUserId) {
          this.handleRemoteEdit(edit);
        }
      });

    this.webSocketService.subscribe<CollaborationActivity>(`/topic/dossier/${dossierId}/activity`)
      .subscribe(activity => {
        this.activity$.next(activity);
      });

    this.webSocketService.subscribe<ConflictResolution>(`/topic/dossier/${dossierId}/conflict`)
      .subscribe(conflict => {
        this.conflict$.next(conflict);
      });

    this.webSocketService.subscribe<SharedFilterPreset>(`/topic/filter-presets`)
      .subscribe(filterPreset => {
        this.filterPreset$.next(filterPreset);
      });

    this.webSocketService.subscribe<SharedFilterPreset>(`/user/${this.currentUserId}/queue/filter-presets`)
      .subscribe(filterPreset => {
        this.filterPreset$.next(filterPreset);
      });
  }

  private joinDossier(dossierId: number, userId: string, username: string): void {
    this.webSocketService.send(`/app/dossier/${dossierId}/join`, {
      userId,
      username
    });
  }

  leaveDossier(): void {
    if (this.currentDossierId && this.currentUserId && this.currentUsername) {
      this.webSocketService.send(`/app/dossier/${this.currentDossierId}/leave`, {
        userId: this.currentUserId,
        username: this.currentUsername
      });
      this.currentDossierId = null;
    }
  }

  updateCursor(fieldName: string, cursorPosition: number): void {
    if (!this.currentDossierId || !this.currentUserId || !this.currentUsername) return;

    const cursor: CollaborationCursor = {
      userId: this.currentUserId,
      username: this.currentUsername,
      dossierId: this.currentDossierId,
      fieldName,
      cursorPosition,
      color: '',
      timestamp: new Date()
    };

    this.webSocketService.send(`/app/dossier/${this.currentDossierId}/cursor`, cursor);
  }

  broadcastEdit(fieldName: string, newValue: any, oldValue: any): CollaborationEdit {
    if (!this.currentDossierId || !this.currentUserId || !this.currentUsername) {
      throw new Error('Collaboration not initialized');
    }

    const version = (this.fieldVersions.get(fieldName) || 0) + 1;
    this.fieldVersions.set(fieldName, version);

    const edit: CollaborationEdit = {
      userId: this.currentUserId,
      username: this.currentUsername,
      dossierId: this.currentDossierId,
      fieldName,
      newValue,
      oldValue,
      version,
      editId: this.generateEditId(),
      timestamp: new Date()
    };

    this.pendingEdits.set(edit.editId, edit);
    this.webSocketService.send(`/app/dossier/${this.currentDossierId}/edit`, edit);

    return edit;
  }

  confirmEdit(editId: string): void {
    this.pendingEdits.delete(editId);
  }

  private handleRemoteEdit(edit: CollaborationEdit): void {
    const localVersion = this.fieldVersions.get(edit.fieldName) || 0;
    
    if (edit.version > localVersion) {
      this.fieldVersions.set(edit.fieldName, edit.version);
      this.edit$.next(edit);
    } else {
      console.warn('Conflict detected for field:', edit.fieldName);
    }
  }

  broadcastActivity(activityType: string, description: string, data: any): void {
    if (!this.currentDossierId || !this.currentUserId || !this.currentUsername) return;

    const activity: CollaborationActivity = {
      userId: this.currentUserId,
      username: this.currentUsername,
      dossierId: this.currentDossierId,
      activityType,
      description,
      data,
      timestamp: new Date()
    };

    this.webSocketService.send(`/app/dossier/${this.currentDossierId}/activity`, activity);
  }

  shareFilterPreset(filterPreset: SharedFilterPreset): void {
    this.webSocketService.send('/app/filter-preset/share', filterPreset);
  }

  resolveConflict(fieldName: string, resolvedValue: any): Observable<void> {
    if (!this.currentDossierId || !this.currentUserId) {
      throw new Error('Collaboration not initialized');
    }

    return this.http.post<void>(
      `/api/collaboration/dossier/${this.currentDossierId}/conflict/resolve`,
      {
        fieldName,
        resolvedValue,
        resolvedBy: this.currentUserId
      }
    );
  }

  getViewers(): Observable<Set<string>> {
    return this.viewers$.asObservable();
  }

  getPresenceUpdates(): Observable<CollaborationPresence> {
    return this.presence$.asObservable();
  }

  getCursorUpdates(): Observable<CollaborationCursor> {
    return this.cursor$.asObservable();
  }

  getEditUpdates(): Observable<CollaborationEdit> {
    return this.edit$.asObservable();
  }

  getActivityUpdates(): Observable<CollaborationActivity> {
    return this.activity$.asObservable();
  }

  getConflictUpdates(): Observable<ConflictResolution> {
    return this.conflict$.asObservable();
  }

  getFilterPresetUpdates(): Observable<SharedFilterPreset> {
    return this.filterPreset$.asObservable();
  }

  getCurrentViewers(dossierId: number): Observable<Set<string>> {
    return this.http.get<Set<string>>(`/api/collaboration/dossier/${dossierId}/viewers`);
  }

  getCurrentVersion(dossierId: number): Observable<{ dossierId: number; version: number }> {
    return this.http.get<{ dossierId: number; version: number }>(
      `/api/collaboration/dossier/${dossierId}/version`
    );
  }

  getUserColor(userId: string): Observable<{ userId: string; color: string }> {
    return this.http.get<{ userId: string; color: string }>(
      `/api/collaboration/user/${userId}/color`
    );
  }

  private generateEditId(): string {
    return `${this.currentUserId}-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  ngOnDestroy(): void {
    this.leaveDossier();
    this.webSocketService.disconnect();
  }
}
